File: apollo-integration/src/test/java/com/apollographql/apollo/ApolloLiveDataTest.java
package com.apollographql.apollo;

import android.arch.core.executor.testing.InstantTaskExecutorRule;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;

import com.apollographql.apollo.api.Input;
import com.apollographql.apollo.cache.normalized.lru.EvictionPolicy;
import com.apollographql.apollo.cache.normalized.lru.LruNormalizedCacheFactory;
import com.apollographql.apollo.integration.normalizer.EpisodeHeroNameQuery;
import com.apollographql.apollo.integration.normalizer.type.Episode;
import com.apollographql.apollo.livedata.LiveDataApollo;
import com.apollographql.apollo.livedata.LiveDataResponse;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;

import okhttp3.Dispatcher;
import okhttp3.OkHttpClient;
import okhttp3.mockwebserver.MockWebServer;

import static com.apollographql.apollo.Utils.immediateExecutor;
import static com.apollographql.apollo.Utils.immediateExecutorService;
import static com.apollographql.apollo.Utils.mockResponse;
import static com.apollographql.apollo.fetcher.ApolloResponseFetchers.NETWORK_ONLY;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsInstanceOf.instanceOf;
import static org.junit.Assert.assertNotNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;

@SuppressWarnings({"unchecked", "unused"})
public class ApolloLiveDataTest {
    private ApolloClient apolloClient;
    private MockWebServer mockWebServer = new MockWebServer();

    private static final String FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json";
    private static final String FILE_EPISODE_HERO_NAME_CHANGE = "EpisodeHeroNameResponseNameChange.json";
    private static final String FILE_HERO_AND_FRIEND_NAME_CHANGE = "HeroAndFriendsNameWithIdsNameChange.json";

    @Rule public final InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    @Before
    public void setup() throws IOException {
        mockWebServer.start();

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .dispatcher(new Dispatcher(immediateExecutorService()))
                .build();

        apolloClient = ApolloClient.builder()
                .serverUrl(mockWebServer.url("/"))
                .dispatcher(immediateExecutor())
                .okHttpClient(okHttpClient)
                .normalizedCache(new LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION), new IdFieldCacheKeyResolver())
                .build();
    }

    @After
    public void stopServer() throws IOException {
        mockWebServer.shutdown();
    }

    @Test
    public void callProducesValue() throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))));
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Success.class));

        LiveDataResponse.Success success = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data = (EpisodeHeroNameQuery.Data) success.getData();
        assertThat(data.hero().__typename() , is("Droid"));
        assertThat(data.hero().name() , is("R2-D2"));
    }

    @Test
    public void prefetchCompletes() throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.prefetch(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))));
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Complete.class));
    }

    @Test
    public void queryWatcherUpdatedSameQueryDifferentResults() throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))).watcher());
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Success.class));

        LiveDataResponse.Success success = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data = (EpisodeHeroNameQuery.Data) success.getData();
        assertThat(data.hero().__typename() , is("Droid"));
        assertThat(data.hero().name() , is("R2-D2"));

        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE));
        apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
                .responseFetcher(NETWORK_ONLY)
                .enqueue(null);

        LiveDataResponse.Success success2 = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data2 = (EpisodeHeroNameQuery.Data) success2.getData();
        assertThat(data2.hero().__typename() , is("Droid"));
        assertThat(data2.hero().name() , is("Artoo"));
    }

    @Test
    public void queryWatcherNotUpdatedSameQuerySameResults() throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))).watcher());
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Success.class));

        LiveDataResponse.Success success = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data = (EpisodeHeroNameQuery.Data) success.getData();
        assertThat(data.hero().__typename() , is("Droid"));
        assertThat(data.hero().name() , is("R2-D2"));

        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE));
        apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
                .responseFetcher(NETWORK_ONLY)
                .enqueue(null);

        verifyNoMoreInteractions(observer);

        LiveDataResponse.Success success2 = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data2 = (EpisodeHeroNameQuery.Data) success2.getData();
        assertThat(data2.hero().__typename() , is("Droid"));
        assertThat(data2.hero().name() , is("R2-D2"));
    }

    @Test
    public void queryWatcherUpdatedDifferentQueryDifferentResults(LiveDataResponse.Success value) throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))).watcher());
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Success.class));

        LiveDataResponse.Success success = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data = (EpisodeHeroNameQuery.Data) success.getData();
        assertThat(data.hero().__typename() , is("Droid"));
        assertThat(data.hero().name() , is("R2-D2"));

        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE));
        apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
                .responseFetcher(NETWORK_ONLY)
                .enqueue(null);

        LiveDataResponse.Success success2 = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data2 = (EpisodeHeroNameQuery.Data) success2.getData();
        assertThat(data2.hero().__typename() , is("Droid"));
        assertThat(data2.hero().name() , is("Artoo"));
    }

    @Test
    public void queryWatcherNotCalledWhenCanceled(LiveDataResponse.Success value) throws Exception {
        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID));

        Observer<LiveDataResponse<EpisodeHeroNameQuery.Data>> observer = mock(Observer.class);
        LiveData liveData = LiveDataApollo.INSTANCE.from(apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE))).watcher());
        liveData.observeForever(observer);

        assertNotNull(liveData.getValue());
        verify(observer).onChanged((LiveDataResponse<EpisodeHeroNameQuery.Data>) (liveData.getValue()));
        assertThat(liveData.getValue(), instanceOf(LiveDataResponse.Success.class));

        LiveDataResponse.Success success = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data = (EpisodeHeroNameQuery.Data) success.getData();
        assertThat(data.hero().__typename() , is("Droid"));
        assertThat(data.hero().name() , is("R2-D2"));

        liveData.removeObserver(observer);

        mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE));
        apolloClient.query(new EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
                .responseFetcher(NETWORK_ONLY)
                .enqueue(null);

        verifyNoMoreInteractions(observer);

        LiveDataResponse.Success success2 = (LiveDataResponse.Success) liveData.getValue();
        EpisodeHeroNameQuery.Data data2 = (EpisodeHeroNameQuery.Data) success2.getData();
        assertThat(data2.hero().__typename() , is("Droid"));
        assertThat(data2.hero().name() , is("R2-D2"));
    }
}

File: apollo-integration/src/test/java/com/apollographql/apollo/ApolloLiveDataTest.kt
package com.apollographql.apollo


import android.arch.core.executor.ArchTaskExecutor
import android.arch.core.executor.TaskExecutor
import android.arch.lifecycle.Observer
import com.apollographql.apollo.Utils.immediateExecutor
import com.apollographql.apollo.Utils.immediateExecutorService
import com.apollographql.apollo.Utils.mockResponse
import com.apollographql.apollo.api.Input
import com.apollographql.apollo.cache.normalized.lru.EvictionPolicy
import com.apollographql.apollo.cache.normalized.lru.LruNormalizedCacheFactory
import com.apollographql.apollo.fetcher.ApolloResponseFetchers.NETWORK_ONLY
import com.apollographql.apollo.integration.normalizer.EpisodeHeroNameQuery
import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesWithIDsQuery
import com.apollographql.apollo.integration.normalizer.type.Episode
import com.apollographql.apollo.livedata.ApolloLiveDataResponse
import com.apollographql.apollo.livedata.toLiveData
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import okhttp3.Dispatcher
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockWebServer
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.instanceOf
import org.hamcrest.MatcherAssert.assertThat
import org.junit.After
import org.junit.Assert.assertNotNull
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.io.IOException

@RunWith(JUnit4::class)
class ApolloLiveDataTest {
  private lateinit var apolloClient: ApolloClient
  private lateinit var mockWebServer: MockWebServer

  private val FILE_EPISODE_HERO_NAME_WITH_ID = "EpisodeHeroNameResponseWithId.json"
  private val FILE_EPISODE_HERO_NAME_CHANGE = "EpisodeHeroNameResponseNameChange.json"
  private val FILE_HERO_AND_FRIEND_NAMD_CHANGE = "HeroAndFriendsNameWithIdsNameChange.json"

  @Before
  @Throws(IOException::class)
  fun setup() {
    mockWebServer = MockWebServer()
    mockWebServer.start()

    val okHttpClient = OkHttpClient.Builder()
        .dispatcher(Dispatcher(immediateExecutorService()))
        .build()

    apolloClient = ApolloClient.builder()
        .serverUrl(mockWebServer.url("/"))
        .dispatcher(immediateExecutor())
        .okHttpClient(okHttpClient)
        .normalizedCache(LruNormalizedCacheFactory(EvictionPolicy.NO_EVICTION), IdFieldCacheKeyResolver())
        .build()

    ArchTaskExecutor.getInstance().setDelegate(object : TaskExecutor() {
      override fun executeOnDiskIO(runnable: Runnable) = runnable.run()
      override fun isMainThread() = true
      override fun postToMainThread(runnable: Runnable) = runnable.run()
    })
  }

  @After
  @Throws(IOException::class)
  fun stopServer() {
    mockWebServer.shutdown()
    ArchTaskExecutor.getInstance().setDelegate(null)
  }

  @Test
  @Throws(Exception::class)
  fun callProducesValue() {
    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))

    val response = LiveDataTestUtil.getValue(apolloClient
        .query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .toLiveData())
    assertThat(response, instanceOf(ApolloLiveDataResponse.Success::class.java))

    val success = response as ApolloLiveDataResponse.Success
    assertThat(success.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success.data?.hero()?.name(), `is`("R2-D2"))
  }

  @Test
  @Throws(Exception::class)
  fun prefetchCompletes() {
    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))

    val response = LiveDataTestUtil.getValue(apolloClient
        .prefetch(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .toLiveData<EpisodeHeroNameQuery.Data>())
    assertThat(response, instanceOf(ApolloLiveDataResponse.Complete::class.java))
  }

  @Test
  @Throws(Exception::class)
  fun queryWatcherUpdatedSameQueryDifferentResults() {
    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))

    val observer = mock<Observer<ApolloLiveDataResponse<EpisodeHeroNameQuery.Data>>>()
    val liveData = apolloClient
        .query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .watcher()
        .toLiveData()
    liveData.observeForever(observer)

    assertNotNull(liveData.value)
    verify(observer).onChanged(liveData.value)
    assertThat(liveData.value, instanceOf(ApolloLiveDataResponse.Success::class.java))

    val success = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success.data?.hero()?.name(), `is`("R2-D2"))

    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_CHANGE))
    apolloClient.query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .responseFetcher(NETWORK_ONLY)
        .enqueue(null)

    val success2 = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success2.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success2.data?.hero()?.name(), `is`("Artoo"))
  }

  @Test
  @Throws(Exception::class)
  fun queryWatcherNotUpdatedSameQuerySameResults() {
    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))

    val observer = mock<Observer<ApolloLiveDataResponse<EpisodeHeroNameQuery.Data>>>()
    val liveData = apolloClient
        .query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .watcher()
        .toLiveData()
    liveData.observeForever(observer)

    assertNotNull(liveData.value)
    verify(observer).onChanged(liveData.value)
    assertThat(liveData.value, instanceOf(ApolloLiveDataResponse.Success::class.java))

    val success = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success.data?.hero()?.name(), `is`("R2-D2"))

    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))
    apolloClient.query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .responseFetcher(NETWORK_ONLY)
        .enqueue(null)

    verifyNoMoreInteractions(observer)

    val success2 = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success2.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success2.data?.hero()?.name(), `is`("R2-D2"))
  }

  @Test
  @Throws(Exception::class)
  fun queryWatcherUpdatedDifferentQueryDifferentResults() {
    mockWebServer.enqueue(mockResponse(FILE_EPISODE_HERO_NAME_WITH_ID))

    val observer = mock<Observer<ApolloLiveDataResponse<EpisodeHeroNameQuery.Data>>>()
    val liveData = apolloClient
        .query(EpisodeHeroNameQuery(Input.fromNullable(Episode.EMPIRE)))
        .watcher()
        .toLiveData()
    liveData.observeForever(observer)

    assertNotNull(liveData.value)
    verify(observer).onChanged(liveData.value)
    assertThat(liveData.value, instanceOf(ApolloLiveDataResponse.Success::class.java))

    val success = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success.data?.hero()?.name(), `is`("R2-D2"))

    mockWebServer.enqueue(mockResponse(FILE_HERO_AND_FRIEND_NAMD_CHANGE))
    apolloClient
        .query(HeroAndFriendsNamesWithIDsQuery(Input.fromNullable(Episode.NEWHOPE)))
        .enqueue(null)

    val success2 = liveData.value as ApolloLiveDataResponse.Success
    assertThat(success2.data?.hero()?.__typename(), `is`("Droid"))
    assertThat(success2.data?.hero()?.name(), `is`("Artoo"))
  }
}